[SaferCpp] InjectedBundle*Handle.cpp

kmonsen opened the pull request:
#### 3d8fd44714d76d420d677fb5f55ed3434b4e819d
<pre>
[SaferCpp] InjectedBundle*Handle.cpp
<a href="https://rdar.apple.com/149529249">rdar://149529249</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291730">https://bugs.webkit.org/show_bug.cgi?id=291730</a>

Reviewed by Rupin Mittal.

* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::protectedCellAbove const):
* Source/WebCore/html/HTMLTableCellElement.h:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
(WebKit::InjectedBundleNodeHandle::getOrCreate):
(WebKit::InjectedBundleNodeHandle::~InjectedBundleNodeHandle):
(WebKit::InjectedBundleNodeHandle::protectedCoreNode):
(WebKit::InjectedBundleNodeHandle::document):
(WebKit::InjectedBundleNodeHandle::absoluteBoundingRect):
(WebKit::InjectedBundleNodeHandle::renderedImage):
(WebKit::InjectedBundleNodeHandle::htmlTableCellElementCellAbove):
(WebKit::InjectedBundleNodeHandle::documentFrame):
(WebKit::InjectedBundleNodeHandle::htmlIFrameElementContentFrame):
(WebKit::InjectedBundleNodeHandle::stop):
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.h:
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleRangeHandle.cpp:
(WebKit::InjectedBundleRangeHandle::getOrCreate):
(WebKit::InjectedBundleRangeHandle::document):
(WebKit::InjectedBundleRangeHandle::boundingRectInWindowCoordinates const):
(WebKit::InjectedBundleRangeHandle::renderedImage):

Canonical link: <a href="https://commits.webkit.org/294032@main">https://commits.webkit.org/294032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab0cd6d66ddf4c256e5210fe70f4686b078c2d19

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100619 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20271 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105756 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51207 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102660 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20579 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28745 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76621 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33662 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90885 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56976 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8892 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50583 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85520 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8967 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108110 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20373 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85575 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28100 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87086 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85115 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21654 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29813 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7539 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21725 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27672 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32923 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27483 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30801 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29041 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
```diff
@@ -89,12 +89,12 @@ RefPtr<InjectedBundleNodeHandle> InjectedBundleNodeHandle::getOrCreate(Node* nod
     auto nodeHandle = InjectedBundleNodeHandle::create(node);
     if (nodeHandle->coreNode())
-        domNodeHandleCache().add(*nodeHandle->coreNode(), nodeHandle.get());
+        domNodeHandleCache().add(*nodeHandle->protectedCoreNode(), nodeHandle.get());
```
Perhaps we can do this to avoid an extra function call:

`if (RefPtr codeNode = nodeHandle->coreNode())`
`       domNodeHandleCache().add(*coreNode, nodeHandle.get());`

kmonsen replied:
I had to alter it a little bit to:
```
if (RefPtr node = nodeHandle->coreNode())
       domNodeHandleCache().add(*node, nodeHandle.get());
```
Because `coreNode` clashes with a member function and the compiler didn't like that

RupinMittal commented on Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
```diff
@@ -114,20 +114,25 @@ InjectedBundleNodeHandle::InjectedBundleNodeHandle(Node& node)
 {
     if (m_node)
-        domNodeHandleCache().remove(*m_node);
+        domNodeHandleCache().remove(*protectedCoreNode());
```
Could avoid an extra function call with:

`if (RefPtr node = m_node)`
`       domNodeHandleCache().remove(*node);`

RupinMittal commented on Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
```diff
@@ -228,7 +233,7 @@ RefPtr<WebImage> InjectedBundleNodeHandle::renderedImage(SnapshotOptions options
 
     frameView->setNodeToDraw(m_node.get());
-    auto image = imageForRect(frameView, paintingRect, bitmapWidth, options);
+    auto image = imageForRect(frameView.get(), paintingRect, bitmapWidth, options);
```
This doesn't pertain to making the code safer (or the rest of your patch), but I believe we prefer to not hide smart pointers behind `auto` where possible. Maybe we can do:

`RefPtr image = imageForRect(frameView.get(), paintingRect, bitmapWidth, options);`

as a drive-by fix here since `imageForRect()` returns a `RefPtr`.

RupinMittal commented on Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
```diff
@@ -460,7 +465,7 @@ void InjectedBundleNodeHandle::stop()
     if (m_node) {
-        domNodeHandleCache().remove(*m_node);
+        domNodeHandleCache().remove(*protectedCoreNode());
         m_node = nullptr;
```
Maybe:

`if (RefPtr node = m_node) {`
`      domNodeHandleCache().remove(*node);`
`      m_node = nullptr;`
`}`

RupinMittal reviewed (approved):
Just a couple nits. They are mostly stylistic suggestions that would be nice to have (based on what I've seen in many other smart pointer patches and my own). I'll leave it up to you. Approved :)

kmonsen committed:
[SaferCpp] InjectedBundle*Handle.cpp
rdar://149529249
https://bugs.webkit.org/show_bug.cgi?id=291730

Reviewed by Rupin Mittal.

* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::protectedCellAbove const):
* Source/WebCore/html/HTMLTableCellElement.h:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
(WebKit::InjectedBundleNodeHandle::getOrCreate):
(WebKit::InjectedBundleNodeHandle::~InjectedBundleNodeHandle):
(WebKit::InjectedBundleNodeHandle::protectedCoreNode):
(WebKit::InjectedBundleNodeHandle::document):
(WebKit::InjectedBundleNodeHandle::absoluteBoundingRect):
(WebKit::InjectedBundleNodeHandle::renderedImage):
(WebKit::InjectedBundleNodeHandle::htmlTableCellElementCellAbove):
(WebKit::InjectedBundleNodeHandle::documentFrame):
(WebKit::InjectedBundleNodeHandle::htmlIFrameElementContentFrame):
(WebKit::InjectedBundleNodeHandle::stop):
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.h:
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleRangeHandle.cpp:
(WebKit::InjectedBundleRangeHandle::getOrCreate):
(WebKit::InjectedBundleRangeHandle::document):
(WebKit::InjectedBundleRangeHandle::boundingRectInWindowCoordinates const):
(WebKit::InjectedBundleRangeHandle::renderedImage):

Canonical link: https://commits.webkit.org/294032@main
